1-(3) 문제

In [8]:
import numpy as np

p = 0.1

def function(x_coordinate):
    a = 2*(x_coordinate[0]**2) + 3*x_coordinate[0]*x_coordinate[1] 
    b = 2*(x_coordinate[1]**2) - 4*x_coordinate[0] +2*x_coordinate[1] -24
    return  a+b


def gradient(x_coordinate):
    a = 4*x_coordinate[0] + 3*x_coordinate[1] - 4
    b = 3*x_coordinate[0] + 4*x_coordinate[1] + 2
    return a, b

x_0 = np.array([1.0, 0.9])
f_x0 = function(x_0)
gradient_x0 = np.array(gradient(x_0))

x_list = [x_0]
for i in range(50):
    grad = np.array(gradient(x_list[i]))
    x_list.append(x_list[i] - p*grad)
    if i <= 3 or i%5 == 0:
        print()
        print("idx: {}, [x1, x2]: {}".format(i, x_list[i]))
        print("f(x1, x2): {}".format(function(x_list[i])))
        print("grad({}): {}, p*grad: {}".format(i,grad, p*grad))


idx: 0, [x1, x2]: [1.  0.9]
f(x1, x2): -19.88
grad(0): [2.7 8.6], p*grad: [0.27 0.86]

idx: 1, [x1, x2]: [0.73 0.04]
f(x1, x2): -25.6834
grad(1): [-0.96  4.35], p*grad: [-0.096  0.435]

idx: 2, [x1, x2]: [ 0.826 -0.395]
f(x1, x2): -27.396208
grad(2): [-1.881  2.898], p*grad: [-0.1881  0.2898]

idx: 3, [x1, x2]: [ 1.0141 -0.6848]
f(x1, x2): -28.514667340000003
grad(3): [-1.998   2.3031], p*grad: [-0.1998   0.23031]

idx: 5, [x1, x2]: [ 1.402873 -1.113236]
f(x1, x2): -30.108456089434
grad(5): [-1.728216  1.755675], p*grad: [-0.1728216  0.1755675]

idx: 10, [x1, x2]: [ 2.11426051 -1.82853669]
f(x1, x2): -32.084836306123705
grad(10): [-1.02856804  1.02863476], p*grad: [-0.1028568   0.10286348]

idx: 15, [x1, x2]: [ 2.53547831 -2.24976401]
f(x1, x2): -32.77394808796707
grad(15): [-0.60737876  0.60737892], p*grad: [-0.06073788  0.06073789]

idx: 20, [x1, x2]: [ 2.78420601 -2.49849173]
f(x1, x2): -33.014226509059306
grad(20): [-0.35865113  0.35865113], p*grad: [-0.03586511  0.03586511]

idx:

x1 = (0.73, 0.04), x2 = (0.826, -0.395), x3 = (1.0141, -0.6848)로 점차 변하고 좀더 연속적으로 계산을 하면 위의 (2)에서 구한 최소점으로 가고 점차 실제 함수 값도 최솟값으로 가는 경향을 확인을 할 수 있다.